packages/atproto_client/models/string_formats.py

```diff
@@ -0,0 +1,154 @@
+import re
+import string
+from datetime import datetime
+from typing import Callable, Mapping, Set, Union
+from urllib.parse import urlparse
+
+from pydantic import BeforeValidator, Field, ValidationInfo
+from pydantic_core import core_schema
+from typing_extensions import Annotated, Literal
+
+_OPT_IN_KEY: Literal['strict_string_format'] = 'strict_string_format'
+
+# constants
+MAX_HANDLE_LENGTH: int = 253
+MAX_NSID_LENGTH: int = 317
+MAX_RECORD_KEY_LENGTH: int = 512
+MAX_URI_LENGTH: int = 8 * 1024
+MIN_CID_LENGTH: int = 8
+TID_LENGTH: int = 13
+INVALID_RECORD_KEYS: Set[str] = {'.', '..'}
+
+# patterns
+DOMAIN_RE = re.compile(r'([a-z0-9][a-z0-9-]{0,62}(?<!-)\.){1,}[a-z][a-z0-9-]*(?<!-)')
+DID_RE = re.compile(r'did:[a-z]+:[A-Za-z0-9._%:-]{1,2048}(?<!:)')
+NSID_RE = re.compile(r'(?![0-9])((?!-)[a-z0-9-]{1,63}(?<!-)\.){2,}[a-zA-Z]{1,63}')
+LANG_RE = re.compile(r'^(i|[a-z]{2,3})(-[A-Za-z0-9-]+)?$')
+RKEY_RE = re.compile(r'^[A-Za-z0-9._:~-]{1,512}$')
+TID_RE = re.compile(rf'^[{string.ascii_lowercase}234567]{{{TID_LENGTH}}}$')
+CID_RE = re.compile(r'^[A-Za-z0-9+]{8,}$')
+AT_URI_RE = re.compile(r'at://[^/]+(/[^/]+(/[^/]+)?)?')
+
+
+def only_validate_if_strict(validate_fn: core_schema.WithInfoValidatorFunction) -> Callable:
+    """Skip validation if not opting into strict validation."""
+
+    def wrapper(v: str, info: ValidationInfo) -> str:
+        if not (info and isinstance(info.context, Mapping) and info.context.get(_OPT_IN_KEY, False)):
+            return v
+        return validate_fn(v, info)
+
+    return wrapper
```

Comment:
here's the critical "pydantic usage" making use of [validation context](https://docs.pydantic.dev/latest/concepts/validators/#validation-context)

Reply:
could we try to invert this if statement? to avoid using of "not"

smth like:

```suggestion
def only_validate_if_strict(validate_fn: core_schema.WithInfoValidatorFunction) -> Callable:
    """Skip validation if not opting into strict validation."""

    def wrapper(v: str, info: ValidationInfo) -> str:
        if info and isinstance(info.context, Mapping) and info.context.get(_OPT_IN_KEY, False):
            return validate_fn(v, info)
        return v

    return wrapper
```

Reply:
aha, good call. updated